key-share/src/trusted_dealer.rs

```diff
@@ -97,18 +97,73 @@ impl<E: Curve> TrustedDealerBuilder<E> {
     ///
     /// Returns error if provided inputs are invalid, or if internal
     /// error has occurred.
+    ///
+    /// For Shamir secret sharing, it's shared at points `1` to `n`
+    ///
+    /// Returns error if provided inputs are invalid, or if internal
+    /// error has occurred.
     pub fn generate_shares(
         self,
         rng: &mut (impl rand_core::RngCore + rand_core::CryptoRng),
     ) -> Result<Vec<CoreKeyShare<E>>, TrustedDealerError> {
-        let shared_secret_key = self
-            .shared_secret_key
-            .unwrap_or_else(|| NonZero::<SecretScalar<_>>::random(rng));
-        let shared_public_key = Point::generator() * &shared_secret_key;
         let key_shares_indexes = (1..=self.n)
             .map(|i| generic_ec::NonZero::from_scalar(Scalar::from(i)))
             .collect::<Option<Vec<_>>>()
             .ok_or(Reason::DeriveKeyShareIndex)?;
+        self.generate_shares_at(key_shares_indexes, rng)
+    }
+
+    /// Generates [`CoreKeyShare`]s shared at random points
+    ///
+    /// Returns error if provided inputs are invalid, or if internal
+    /// error has occurred.
+    ///
+    /// For Shamir secret sharing, the points at which the value is shared at
+    /// are chosen at random between `1` and `u16::MAX`. For additive shares,
+    /// this is the same as [`TrustedDealerBuilder::generate_shares`]
+    ///
+    /// Returns error if provided inputs are invalid, or if internal
+    /// error has occurred.
+    pub fn generate_shares_at_random(
+        self,
+        rng: &mut (impl rand_core::RngCore + rand_core::CryptoRng),
+    ) -> Result<Vec<CoreKeyShare<E>>, TrustedDealerError> {
+        let mut points = Vec::with_capacity(self.n.into());
+        'each_point: for _ in 0..self.n {
+            for _ in 0..u16::MAX {
+                let point = generic_ec::NonZero::<Scalar<E>>::random(rng);
+                if !points.contains(&point) {
+                    points.push(point);
+                    continue 'each_point;
+                }
+            }
+            // if we did not continue in inner loop, it means we couldn't
+            // generate a distinct scalar
+            return Err(Reason::BadRandom.into());
+        }
+        self.generate_shares_at(points, rng)
+    }
+
+    /// Generates [`CoreKeyShare`]s shared at preimages provided. Each share is
+    /// going to have the given `preimages` as its `I` component.
+    ///
+    /// Preimages are ignored for additive key shares.
+    ///
+    /// Returns error if provided inputs are invalid, or if internal
+    /// error has occurred.
+    pub fn generate_shares_at(
+        self,
+        preimages: Vec<NonZero<Scalar<E>>>,
+        rng: &mut (impl rand_core::RngCore + rand_core::CryptoRng),
+    ) -> Result<Vec<CoreKeyShare<E>>, TrustedDealerError> {
```

Comment:
In this function it might make sense to add a check that all preimages are distinct to protect from invalid input, but I think the key share validation logic already does that, so no need to do that here

Reply:
It doesn't check it explicitly, but if we have duplicate preimages, we will divide by zero in one of lagrange interpolations, which will be caught